File: backend/services/auth.service.ts
/**
 * Auth Service — JWT via HTTP-only cookies + bcrypt password hashing
 */
import bcrypt from 'bcryptjs';
import { SignJWT, jwtVerify } from 'jose';
import { env } from '../config/env.js';

const secret = new TextEncoder().encode(env.JWT_SECRET);
const SALT_ROUNDS = 12;

export async function hashPassword(password: string): Promise<string> {
  return bcrypt.hash(password, SALT_ROUNDS);
}

export async function verifyPassword(password: string, hash: string): Promise<boolean> {
  return bcrypt.compare(password, hash);
}

export interface JwtPayload {
  sub: string;    // user id
  email: string;
  role: 'assureur' | 'assure';
}

export async function signToken(payload: JwtPayload): Promise<string> {
  return new SignJWT({ ...payload })
    .setProtectedHeader({ alg: 'HS256' })
    .setIssuedAt()
    .setExpirationTime(env.JWT_EXPIRATION)
    .sign(secret);
}

export async function verifyToken(token: string): Promise<JwtPayload | null> {
  try {
    const { payload } = await jwtVerify(token, secret);
    return {
      sub: payload.sub as string,
      email: payload['email'] as string,
      role: payload['role'] as 'assureur' | 'assure',
    };
  } catch {
    return null;
  }
}

export function buildCookieHeader(token: string): string {
  const maxAge = 24 * 60 * 60; // 24h in seconds
  const secure = env.COOKIE_SECURE ? '; Secure' : '';
  return `token=${token}; HttpOnly; SameSite=Lax; Path=/; Max-Age=${maxAge}${secure}`;
}

export function clearCookieHeader(): string {
  return `token=; HttpOnly; SameSite=Lax; Path=/; Max-Age=0`;
}

File: backend/services/auth_service.py
"""
Auth Service — JWT via httpOnly cookies + bcrypt password hashing
=================================================================
Direct port of `backend/services/auth.service.ts`.
"""

from __future__ import annotations

import re

from core.security import (
    JwtPayload,
    build_cookie_value,
    clear_cookie_value,
    create_access_token,
    decode_access_token,
    hash_password,
    verify_password,
)


class AuthService:
    """
    Authentication logic.
    Delegates to `core/security.py` for crypto primitives.
    """

    @staticmethod
    def hash_password(password: str) -> str:
        return hash_password(password)

    @staticmethod
    def verify_password(plain: str, hashed: str) -> bool:
        return verify_password(plain, hashed)

    @staticmethod
    def create_token(user_id: str, email: str, role: str) -> str:
        payload = JwtPayload(sub=user_id, email=email, role=role)  # type: ignore
        return create_access_token(payload)

    @staticmethod
    def verify_token(token: str) -> JwtPayload | None:
        return decode_access_token(token)

    @staticmethod
    def build_cookie(token: str) -> str:
        return build_cookie_value(token)

    @staticmethod
    def clear_cookie() -> str:
        return clear_cookie_value()

    @staticmethod
    def extract_token_from_cookie(cookie_header: str | None) -> str | None:
        """Parse JWT token from the `token` cookie."""
        if not cookie_header:
            return None
        match = re.search(r"(?:^|;\s*)token=([^;]*)", cookie_header)
        return match.group(1) if match else None
